File: python/tests/test_reference_category.py
"""Tests for reference category selection in factor variables and interactions."""

import pytest
import polars as pl
import numpy as np

from leanfe import leanfe_polars, leanfe_duckdb


@pytest.fixture
def region_data():
    """Create test data with region factor."""
    np.random.seed(42)
    n = 1000
    
    df = pl.DataFrame({
        "customer_id": np.repeat(np.arange(100), 10),
        "product_id": np.tile(np.arange(50), 20),
        "region": np.random.choice(["R1", "R2", "R3"], n),
        "treatment": np.random.binomial(1, 0.5, n),
    })
    
    # Generate outcome
    df = df.with_columns([
        (10.0 + 
         pl.col("treatment") * 0.5 +
         pl.col("customer_id").cast(pl.Float64) * 0.01 +
         pl.Series(np.random.normal(0, 1, n))
        ).alias("revenue")
    ])
    
    return df


def test_default_reference_first_polars(region_data):
    """Test that default reference is first category (alphabetically)."""
    result = leanfe_polars(
        region_data,
        formula="revenue ~ treatment + i(region) | customer_id + product_id"
    )
    
    # R1 should be reference (first alphabetically), so we get R2 and R3
    assert "region_R2" in result["coefficients"]
    assert "region_R3" in result["coefficients"]
    assert "region_R1" not in result["coefficients"]


def test_custom_reference_polars(region_data):
    """Test specifying custom reference category in Polars."""
    result = leanfe_polars(
        region_data,
        formula="revenue ~ treatment + i(region, ref=R2) | customer_id + product_id"
    )
    
    # R2 should be reference, so we get R1 and R3
    assert "region_R1" in result["coefficients"]
    assert "region_R3" in result["coefficients"]
    assert "region_R2" not in result["coefficients"]


def test_custom_reference_duckdb(region_data):
    """Test specifying custom reference category in DuckDB."""
    result = leanfe_duckdb(
        region_data,
        formula="revenue ~ treatment + i(region, ref=R3) | customer_id + product_id"
    )
    
    # R3 should be reference, so we get R1 and R2
    assert "region_R1" in result["coefficients"]
    assert "region_R2" in result["coefficients"]
    assert "region_R3" not in result["coefficients"]


def test_interaction_custom_reference_polars(region_data):
    """Test specifying custom reference in interaction term."""
    result = leanfe_polars(
        region_data,
        formula="revenue ~ treatment:i(region, ref=R2) | customer_id + product_id"
    )
    
    # R2 should be reference, so we get treatment_R1 and treatment_R3
    assert "treatment_R1" in result["coefficients"]
    assert "treatment_R3" in result["coefficients"]
    assert "treatment_R2" not in result["coefficients"]


def test_interaction_custom_reference_duckdb(region_data):
    """Test specifying custom reference in interaction term (DuckDB)."""
    result = leanfe_duckdb(
        region_data,
        formula="revenue ~ treatment:i(region, ref=R1) | customer_id + product_id"
    )
    
    # R1 should be reference, so we get treatment_R2 and treatment_R3
    assert "treatment_R2" in result["coefficients"]
    assert "treatment_R3" in result["coefficients"]
    assert "treatment_R1" not in result["coefficients"]


def test_invalid_reference_raises_error(region_data):
    """Test that invalid reference category raises error."""
    with pytest.raises(ValueError, match="Reference category.*not found"):
        leanfe_polars(
            region_data,
            formula="revenue ~ treatment + i(region, ref=INVALID) | customer_id + product_id"
        )


def test_quoted_reference_polars(region_data):
    """Test that quoted reference values work."""
    result = leanfe_polars(
        region_data,
        formula='revenue ~ treatment + i(region, ref="R2") | customer_id + product_id'
    )
    
    # R2 should be reference
    assert "region_R1" in result["coefficients"]
    assert "region_R3" in result["coefficients"]
    assert "region_R2" not in result["coefficients"]

File: r/tests/test_reference_category.R
#!/usr/bin/env Rscript
# Test reference category selection in factor variables and interactions

library(polars)
library(duckdb)
library(DBI)
source("R/common.R")
source("R/compress.R")
source("R/polars.R")
source("R/duckdb.R")
source("R/leanfe.R")

# Generate test data
set.seed(42)
n <- 1000

df <- pl$DataFrame(
  customer_id = rep(1:100, each = 10),
  product_id = rep(1:50, times = 20),
  region = sample(c("R1", "R2", "R3"), n, replace = TRUE),
  treatment = rbinom(n, 1, 0.5)
)

df <- df$with_columns(
  revenue = pl$lit(10.0) +
    pl$col("treatment") * 0.5 +
    pl$col("customer_id")$cast(pl$Float64) * 0.01 +
    pl$lit(rnorm(n, 0, 1))
)

cat("Testing reference category selection...\n\n")

# Test 1: Default reference (first category)
cat("Test 1: Default reference is first category (Polars)... ")
result <- leanfe_polars(df, formula = "revenue ~ treatment + i(region) | customer_id + product_id")
stopifnot("region_R2" %in% names(result$coefficients))
stopifnot("region_R3" %in% names(result$coefficients))
stopifnot(!("region_R1" %in% names(result$coefficients)))  # R1 is reference
cat("PASSED\n")

# Test 2: Custom reference (Polars)
cat("Test 2: Custom reference category (Polars)... ")
result <- leanfe_polars(df, formula = "revenue ~ treatment + i(region, ref=R2) | customer_id + product_id")
stopifnot("region_R1" %in% names(result$coefficients))
stopifnot("region_R3" %in% names(result$coefficients))
stopifnot(!("region_R2" %in% names(result$coefficients)))  # R2 is reference
cat("PASSED\n")

# Test 3: Custom reference (DuckDB)
cat("Test 3: Custom reference category (DuckDB)... ")
result <- leanfe_duckdb(df, formula = "revenue ~ treatment + i(region, ref=R3) | customer_id + product_id")
stopifnot("region_R1" %in% names(result$coefficients))
stopifnot("region_R2" %in% names(result$coefficients))
stopifnot(!("region_R3" %in% names(result$coefficients)))  # R3 is reference
cat("PASSED\n")

# Test 4: Interaction with custom reference (Polars)
cat("Test 4: Interaction with custom reference (Polars)... ")
result <- leanfe_polars(df, formula = "revenue ~ treatment:i(region, ref=R2) | customer_id + product_id")
stopifnot("treatment_R1" %in% names(result$coefficients))
stopifnot("treatment_R3" %in% names(result$coefficients))
stopifnot(!("treatment_R2" %in% names(result$coefficients)))  # R2 is reference
cat("PASSED\n")

# Test 5: Interaction with custom reference (DuckDB)
cat("Test 5: Interaction with custom reference (DuckDB)... ")
result <- leanfe_duckdb(df, formula = "revenue ~ treatment:i(region, ref=R1) | customer_id + product_id")
stopifnot("treatment_R2" %in% names(result$coefficients))
stopifnot("treatment_R3" %in% names(result$coefficients))
stopifnot(!("treatment_R1" %in% names(result$coefficients)))  # R1 is reference
cat("PASSED\n")

# Test 6: Invalid reference raises error
cat("Test 6: Invalid reference raises error... ")
error_caught <- FALSE
tryCatch({
  leanfe_polars(df, formula = "revenue ~ treatment + i(region, ref=INVALID) | customer_id + product_id")
}, error = function(e) {
  if (grepl("not found", e$message)) {
    error_caught <<- TRUE
  }
})
stopifnot(error_caught)
cat("PASSED\n")

# Test 7: Quoted reference values work
cat("Test 7: Quoted reference values... ")
result <- leanfe_polars(df, formula = 'revenue ~ treatment + i(region, ref="R2") | customer_id + product_id')
stopifnot("region_R1" %in% names(result$coefficients))
stopifnot("region_R3" %in% names(result$coefficients))
stopifnot(!("region_R2" %in% names(result$coefficients)))  # R2 is reference
cat("PASSED\n")

cat("\n✅ All 7 tests passed!\n")
